### PR TITLE
feat(scripts): Add dependency version scanner tool

### PR DESCRIPTION
**Pull request by chalmerlowe**
This adds a utility with the ability to scan for common references to dependencies (Python runtimes and package dependencies) to facilitate updating code when runtimes and dependencies change.

* It can be run against an entire repo OR against specific packages within a monorepo
* It is customizable with [regex patterns and examples here](https://github.com/googleapis/google-cloud-python/pull/16867/changes#diff-d17423afcf0604a287af7c6a590da37df7105674bbb715248d4f0c53545f3ea9) 
* The test suite checks each regex against the examples to ensure the efficacy of the patterns
* The current patterns account for edge cases such as finding `< 3.8` when searching for references to `3.7` since they are semantically equivalent even if syntactically different.
* The scanner produces a CSV report with:
```
path/filename, package name, line number, matching pattern, full line for context, etc.
```

### CODE REVIEW

File: scripts/version_scanner/tests/integration/test_scanner_integration.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import csv
```
Comment (daniel-sanche) [inline scripts/version_scanner/tests/integration/test_scanner_integration.py:15]: It looks like the copytight header is missing (applies to all code files)

File: scripts/version_scanner/README.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
+# Automated Dependency Version Scanner
+
+This tool scans the repository for hardcoded references to specific dependency versions (like Python 3.7) that need to be upgraded or removed.
+
+## Usage
+
+Run the script from the repository root:
+
+```bash
+python3 scripts/version_scanner/version_scanner.py -d <dependency> -v <version> [options]
```
Comment (daniel-sanche) [inline scripts/version_scanner/README.md:10]: When I ran this, I gt a ModuleNotFound error. is there a requirements.txt or anything that captures the dependencies?

File: scripts/version_scanner/README.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+## Universal Prompt for EOL Runtime & Dependency Migration
+
+### Context & Overview
+
+#### Overview
+This plan outlines the approach to update Python packages to drop support for end-of-life Python runtimes (3.7, 3.8, 3.9) OR for deprecated dependencies, and ensure the packages are configured for modern Python.
+
+#### High-Level Strategy
+- **One Branch Per Package**: To keep PRs manageable and isolated, we suggest a dedicated worktree and branch for each package (e.g., `feat/drop-<dependency>-<version>-<package-name>` i.e. `feat/drop-protobuf-4.25.8-google-cloud-bigquery`).
```
Comment (daniel-sanche) [inline scripts/version_scanner/README.md:65]: This is only for hand-written packages, right? I assume others would get their updates through the generator?

Should we recommend doing a generator update first, to clean up most of the packages?

File: scripts/version_scanner/small_package_list.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+packages/google-cloud-access-context-manager
```
Comment (daniel-sanche) [inline scripts/version_scanner/small_package_list.txt:2]: what is this?

File: scripts/version_scanner/version_scanner.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+    """Handles loading and interpolation of regex configurations."""
+    
+    def __init__(self, config_path: str, dependency: str, version: str):
+        self.config_path = config_path
+        self.dependency = dependency
+        self.version = version
+        self.variables = self._compute_variables()
+        
+    def _compute_variables(self) -> Dict[str, str]:
+        """Compute variables for interpolation from version string."""
```
Comment (daniel-sanche) [inline scripts/version_scanner/version_scanner.py]: nit: more detailed comments/examples could be helpful for future maintainers. I'm not sure what a variable is, or the expected version string format

File: scripts/version_scanner/version_scanner.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+        
+    Returns:
+        A list of dictionaries containing match details.
+    """
+    results = []
+            
+    try:
+        with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
+            skip_next = False
+            for line_num, line in enumerate(f, 1):
```
Comment (daniel-sanche) [inline scripts/version_scanner/version_scanner.py:165]: are there any issues with statements that span lines?

File: scripts/version_scanner/version_scanner.py [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+                
+        print(f"\nReport written to: {output_path}")
+    except IOError as e:
+        print(f"Error writing CSV report: {e}", file=sys.stderr)
+
+
+def upload_to_drive(csv_path: str, matches: List[Dict[str, str]], github_repo: str = None, branch: str = "main") -> str:
+    """
+    Upload matches to a Google Sheet in Drive.
+    """
```
Comment (daniel-sanche) [inline scripts/version_scanner/version_scanner.py:331]: Is this necessary? It seems to add extra complexity, dependencies and test surface area, when Google Sheets makes it pretty easy to import a csv natively already

File: scripts/version_scanner/version_scanner.py [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+        dirs[:] = [d for d in dirs if d.lower() not in ignore_lower]
+        
+        # Filter ignore files (case-insensitive)
+        files = [f for f in files if f.lower() not in ignore_lower]
+        
+        rel_root = os.path.relpath(root, root_path)
+        parts = rel_root.split(os.sep)
+        
+        # Monorepo filtering
+        if target_packages and parts[0] == "packages":
```
Comment (daniel-sanche) [inline scripts/version_scanner/version_scanner.py]: There's talk of separating the packages directory into separate ones for generated and handwritten libraries. Will that be easy to address here?

File: scripts/version_scanner/version_scanner.py [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+        help="Root directory to scan (defaults to current directory)"
+    )
+    
+
+    
+    package_group = parser.add_mutually_exclusive_group()
+    
+    package_group.add_argument(
+        "--package",
+        help="Specific subdirectory filter (useful for monorepos)"
```
Comment (daniel-sanche) [inline scripts/version_scanner/version_scanner.py:567]: Is this specific to the structure of the monorepo's package directory? Os is this more of a generic subdirectory filter?

File: scripts/version_scanner/tests/integration/test_scanner_integration.py (hunk 1 above)
Comment (chalmerlowe) [reply scripts/version_scanner/tests/integration/test_scanner_integration.py:15]: Added a license header to code files.
Per convention, did not include license header in: 
* config files such as `.gitignore`, `requirements.txt` 
* OR data files used for testing

File: scripts/version_scanner/README.md (hunk 2 above)
Comment (chalmerlowe) [reply scripts/version_scanner/README.md:10]: Added requirements.txt

File: scripts/version_scanner/README.md (hunk 3 above)
Comment (chalmerlowe) [reply scripts/version_scanner/README.md:65]: There is a note to the effect that if the templates in the `gapic-generator` are updated, then the changes will trickle down to generated packages. This note is in the README.md in the vicinity of lines 34 - 38.

File: scripts/version_scanner/small_package_list.txt (hunk 4 above)
Comment (chalmerlowe) [reply scripts/version_scanner/small_package_list.txt:2]: I don't know if you are asking about `google-cloud-access-context-manager` or about the file.

The file `small_package_list.txt` is a way to present a list of packages to the scanner versus scanning only one package OR scanning all packages. The use of this file is explained in the `README.md` near line 19.

The specific packages listed here are just packages chosen at random for example purposes. I asked Gemini to grab a couple package names to help me do a test of functionality.

File: scripts/version_scanner/version_scanner.py (hunk 5 above)
Comment (chalmerlowe) [reply scripts/version_scanner/version_scanner.py]: Added additional comments.

File: scripts/version_scanner/version_scanner.py (hunk 6 above)
Comment (chalmerlowe) [reply scripts/version_scanner/version_scanner.py:165]: Added a limitation in the README.md clarifying that this is solely a **single-line** scanning engine. Mostly because references to version numbers tend to happen on a single line and to keep complexity low. If we determine that not having multi-line versioning is an issue, we can include that as a feature in future updates.

File: scripts/version_scanner/version_scanner.py (hunk 7 above)
Comment (chalmerlowe) [reply scripts/version_scanner/version_scanner.py:331]: Great question! During dependency cleanup sessions, I typically ran the scanner multiple times:
* comparing 'before and after' results
* rescanning a library if I detected a new regex pattern should be used
* and doing manual QA cycles

Having an automated `--upload` feature to instantly publish a shared, readable, sortable and filterable Google Sheet saves significant toil compared to **repeatedly exporting and manually importing** CSVs into Sheets.

File: scripts/version_scanner/version_scanner.py (hunk 8 above)
Comment (chalmerlowe) [reply scripts/version_scanner/version_scanner.py]: We have the ability to handle any aggregating folder from this list:
* `packages/`
* `handwritten/`
* `generated/`
* `third-party/`

other folder names can be added simply enough to the version_scanner if we end up using different naming conventions during a future mono-repo upgrade.

This update can be found near version_scanner.pyL#515

**Review by chalmerlowe — COMMENTED** [review]
@daniel-sanche 

I responded to the comments with some code updates and with some explanations. Please take a look.

File: scripts/version_scanner/version_scanner.py (hunk 9 above)
Comment (chalmerlowe) [reply scripts/version_scanner/version_scanner.py:567]: See note above.

**Review by daniel-sanche — APPROVED** [review]
LGTM
